python_modules/libraries/dagster-mysql/dagster_mysql/event_log/event_log.py

```diff
@@ -200,22 +198,20 @@ def enable_secondary_index(self, name: str) -> None:
     def watch(self, run_id: str, cursor: Optional[str], callback: EventHandlerFn) -> None:
         if cursor and EventLogCursor.parse(cursor).is_offset_cursor():
             check.failed("Cannot call `watch` with an offset cursor")
-        self._event_watcher.watch_run(run_id, cursor, callback)
 
-    def end_watch(self, run_id: str, handler: EventHandlerFn) -> None:
-        self._event_watcher.unwatch_run(run_id, handler)
+        if self._event_watcher is None:
+            self._event_watcher = SqlPollingEventWatcher(self)
 
-    @property
-    def event_watcher(self) -> SqlPollingEventWatcher:
-        return self._event_watcher
+        self._event_watcher.watch_run(run_id, cursor, callback)
 
-    def __del__(self) -> None:
-        self.dispose()
+    def end_watch(self, run_id: str, handler: EventHandlerFn) -> None:
+        if self._event_watcher:
+            self._event_watcher.unwatch_run(run_id, handler)
 
     def dispose(self) -> None:
-        if not self._disposed:
-            self._disposed = True
+        if self._event_watcher:
             self._event_watcher.close()
+            self._event_watcher = None
```

Comment:
is there a way where we can more explicitly control the lifecycle of `SqlPollingEventWatcher`. In the ideal state construction, `end_watch` and setting to `None` would all be in a contextmanager-decorated function.

Reply:
`watch` and `end_watch` get called per run, many times over the course of the existence of a `DagsterInstance` in a webserver process, so I don't think you can get a contextmanager-decorated function to handle the tear down of this one object that manages all watches. 

We could replace `dispose` with `__exit__` if we made this class a context manager but thats not a substantial improvement in my opinion

Reply:
Got it makes sense